src/node_i18n.cc

```diff
@@ -62,6 +72,300 @@ bool flag_icu_data_dir = false;
+
+  UErrorCode status = U_ZERO_ERROR;
+
+  MaybeStackBuffer<char> buf;
+  uint32_t len;
+  uint32_t limit;
+  char* target;
+  const char* data = ts_obj_data;
+
+  Converter to(env, *to_name);
+  Converter from(env, *from_name);
+  if (to.conv == nullptr || from.conv == nullptr)
+    return;
+
+  ucnv_setSubstChars(to.conv, "?", 1, &status);
+  status = U_ZERO_ERROR;
+
+  limit = ts_obj_length * ucnv_getMaxCharSize(to.conv);
+  buf.AllocateSufficientStorage(ts_obj_length * ucnv_getMaxCharSize(to.conv));
+  target = *buf;
+
+  ucnv_convertEx(to.conv, from.conv,
+                 &target, target + limit,
+                 &data, ts_obj_data + ts_obj_length,
+                 nullptr, nullptr, nullptr, nullptr,
+                 true, true,
+                 &status);
+
+  if (U_SUCCESS(status)) {
+    len = target - *buf;
+    args.GetReturnValue().Set(
+        AsBuffer(env->isolate(), &buf, len).ToLocalChecked());
+  } else {
+    ThrowICUError(env, status, "Unable to transcode buffer");
+  }
+}
+
+// Converts to UCS2 from ISO-8859-1 and US-ASCII
+// args[0] is the encoding to convert from
+// args[1] must be a buffer instance
+void ConvertToUcs2(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsString());
+  THROW_AND_RETURN_UNLESS_BUFFER(env, args[1]);
+
+  Utf8Value name(env->isolate(), args[0]);
+
+  SPREAD_BUFFER_ARG(args[1], ts_obj);
+
+  UErrorCode status = U_ZERO_ERROR;
+  size_t len;
+
+  MaybeStackBuffer<UChar> buf(ts_obj_length);
+
+  Converter from(env, *name);
+  if (from.conv == nullptr)
+    return;
+
+  len = ucnv_toUChars(from.conv,
+                      *buf, ts_obj_length * sizeof(*buf),
+                      ts_obj_data, ts_obj_length,
+                      &status);
+
+  if (U_FAILURE(status)) {
+    ThrowICUError(env, status, "Unable to transcode buffer");
+    return;
+  }
+  args.GetReturnValue().Set(
+      AsBuffer(env->isolate(), &buf, ts_obj_length << 1).ToLocalChecked());
+}
+
+// Convert to a named encoding from UCS2
+// args[0] is the name of the encoding being converted to
+// args[1] must be a buffer instance
+void ConvertFromUcs2(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsString());
+  THROW_AND_RETURN_UNLESS_BUFFER(env, args[1]);
+
+  Utf8Value name(env->isolate(), args[0]);
+  SPREAD_BUFFER_ARG(args[1], ts_obj);
+
+  UErrorCode status = U_ZERO_ERROR;
+
+  UChar* source = nullptr;
+  MaybeStackBuffer<UChar> swapspace;
+  const size_t length = ts_obj_length >> 1;
+
+  swapspace.AllocateSufficientStorage(length);
+  if (IsLittleEndian()) {
+    memcpy(*swapspace, ts_obj_data, ts_obj_length);
+  } else {
+    for (size_t n = 0, i = 0; i < length; n += 2, i += 1) {
+      const uint8_t hi = static_cast<uint8_t>(ts_obj_data[n + 0]);
+      const uint8_t lo = static_cast<uint8_t>(ts_obj_data[n + 1]);
+      swapspace[i] = (lo << 8) | hi;
```

Comment:
for future performance enhancement, detect the alignment of the pointer and perform as many swaps that can be done in a single go.